File: Java/649. Dota2 Senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        Queue<Character> q = new LinkedList();
        int rCount = 0, dCount = 0, rBan = 0, dBan = 0;

        for (char c: senate.toCharArray()) {
            q.add(c);
            if (c == 'R') {
                rCount++;
            } else {
                dCount++;
            }
        }

        while (rCount > 0 && dCount > 0) {
            char cur = q.poll();

            if (cur == 'D') {
                if (dBan > 0) {
                    dBan--;
                    dCount--;
                } else {
                    rBan++;
                    q.offer('D');
                }
            } else {
                if (rBan > 0) {
                    rBan--;
                    rCount--;
                } else {
                    dBan++;
                    q.offer('R');
                }
            }
        }

        return rCount > 0 ? "Radiant" : "Dire";
    }
}

File: Python3/649. Dota2 Senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        queue = deque(senate)
        rCount = senate.count('R')
        dCount = len(senate) - rCount

        rBan = dBan = 0

        while rCount and dCount:
            cur = queue.popleft()

            if cur == 'D':
                if dBan:
                    dBan -= 1
                    dCount -= 1

                else:
                    rBan += 1
                    queue.append('D')

            else:
                if rBan:
                    rBan -= 1
                    rCount -= 1

                else:
                    dBan += 1
                    queue.append('R')

        return 'Radiant' if rCount else 'Dire'
            